add rmb diagnostics call

Omarabdul3ziz opened the pull request:
### Description
- add new rmb call `zos.system.diagnostics` returns the output of `zbusdebug` command
- add new perf task to check the node reaches for grid services

### Related Issues
- https://github.com/threefoldtech/zos/issues/2153
- https://github.com/threefoldtech/zos/issues/2163

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstring

Omarabdul3ziz committed:
add rmb diagnostics call

xmonader commented on cmds/modules/noded/main.go:
```diff
@@ -40,6 +41,11 @@ const (
 	eventsBlock     = "/tmp/events.chain"
 )
 
+type Diagnostics struct {
```
Not the right place, should be in client/node.go (next to zos.statistics.get). cmds is only for commands (binaries)

xmonader commented on cmds/modules/noded/main.go:
```diff
@@ -191,6 +197,28 @@ func action(cli *cli.Context) error {
+				return nil, errors.Wrapf(err, "failed to get status of module %s", module)
+			}
+			for _, worker := range status.Workers {
+				if worker.State != "free" {
```
Should compare against `WorkerFree` instead

xmonader commented on cmds/modules/noded/main.go:
```diff
@@ -191,6 +197,28 @@ func action(cli *cli.Context) error {
+			}
+			for _, worker := range status.Workers {
+				if worker.State != "free" {
+					results.SystemStatusOk = false
```
I'm not sure if the comparison against WorkerFree by itself is enough, i think the check should be in status, err := ... and then do if err -> systemOk is false instead (you shouldn't return there)

xmonader commented on cmds/modules/noded/main.go:
```diff
@@ -191,6 +197,28 @@ func action(cli *cli.Context) error {
+
+		for module := range zbusdebug.PossibleModules {
+			status, err := redis.Status(ctx, module)
+			if err != nil {
```
this should flag the systemok as false and not return ...

Omarabdul3ziz committed:
show if error in response,

xmonader commented on cmds/modules/noded/main.go:
```diff
@@ -191,6 +192,42 @@ func action(cli *cli.Context) error {
+
+			if err != nil {
+				results.SystemStatusOk = false
+			} else {
```
no need to check on the workers' status, I double checked with Azmy, being in Free/Busy state doesn't indicate any error

xmonader commented on cmds/modules/noded/main.go:
```diff
@@ -191,6 +192,42 @@ func action(cli *cli.Context) error {
+		}
+
+		for module := range zbusdebug.PossibleModules {
+			status, err := redis.Status(ctx, module)
```
Please make sure the context used is WithTimeout maybe 2 3 seconds

Omarabdul3ziz committed:
remove the free/busy check and add timed-out context

xmonader commented on cmds/modules/noded/main.go:
```diff
@@ -191,6 +192,39 @@ func action(cli *cli.Context) error {
+			Modules:        make(map[string]moduleStatus),
+		}
+
+		for module := range zbusdebug.PossibleModules {
```
Don't depend on zbusdebug, Copy the possible modules into this file or this function please

xmonader replied:
Any reason not to do these module status checks concurrently?

Omarabdul3ziz replied:
just thought having them in one place would be easier if gonna change later. but ok can copy them here

no, not a specific reason but the call takes no time.

xmonader replied:
Worst case scenario if each take 3 seconds to reply and all of them don't and we do sequentially that maybe 3 sec * 7 modules = 21 seconds

muhamadazmy replied:
You definitely should run all the checks at once (make the calls async) and reply on a single response queue or using a waitgroup

xmonader commented on cmds/modules/noded/main.go:
```diff
@@ -191,6 +192,39 @@ func action(cli *cli.Context) error {
+			}
+			results.Modules[module] = moduleStatus
+
+			cancel()
```
cancel should have been deferred, directly after making that context if needed

muhamadazmy commented on cmds/modules/noded/main.go:
```diff
@@ -191,6 +192,39 @@ func action(cli *cli.Context) error {
 		return version, nil
 	})
 
+	bus.WithHandler("zos.system.diagnostics", func(ctx context.Context, payload []byte) (interface{}, error) {
```
i don't like that the implementation is like this in the main file. The WithHandler can still take a function that is at least defined separately as it's own function (not anonymous) and preferably in a separate file

Omarabdul3ziz committed:
separate diagnostics calls in new module

Omarabdul3ziz committed:
add new perf task `network-health` & show the network status on diagnostics call

Omarabdul3ziz committed:
open a tcp socket to check the connectivity of each service

muhamadazmy commented on pkg/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,121 @@
+	Online bool `json:"online"`
+}
+
+func GetSystemDiagnostics(ctx context.Context, msgBrokerCon string) (Diagnostics, error) {
```
it's really bad idea to pass around the msgBrokerCon string instead of the zbus.Client itself. the zbus.Client can be used concurrently by multiple stubs with no issue. Creating a Client creates a pool that is kept alive forever. Hence  creating a client each time means you create a new pool each time u call this function!

muhamadazmy commented on pkg/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,121 @@
+	}
+}
+
+func isOnline(ctx context.Context, msgBrokerCon string) bool {
```
What does this do ?

Omarabdul3ziz replied:
we needed to add an extra check on the diagnosis call to check the node connectivity to the grid services.
so i added a new perf task `network-health` that check services (chain, relay, graphql, ...) 
`isOnline` check if all the services `is_reachable` by getting the perf task result from the redis cache.

muhamadazmy commented on pkg/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	conn := client.Get()
+	defer conn.Close()
```
you closing the connection but not the pool. The idea of a `Pool` is the reusability so creating a pool ONLY to use one connection then through it away does not sound like a good way to use a pool

Instead Pools are created to live long, and used repeatedly to retrieve (and reuse) connections. So at least here you need to close the pool after you re done .

muhamadazmy commented on pkg/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,121 @@
+	conn := client.Get()
+	defer conn.Close()
+
+	data, err := conn.Do("GET", testNetworkKey)
```
Where is that key coming from ? and what are we trying to do here.

Omarabdul3ziz replied:
a constant perf task `perf.network-health`

muhamadazmy commented on pkg/perf/networkhealth/networkhealth.go:
```diff
@@ -0,0 +1,112 @@
+
+	if !strings.Contains(host, ":") {
+		host = fmt.Sprintf("%s:80", host)
+	}
```
The default port should be based on the url `scheme` 80 if it's http, and 443 if it's https

Omarabdul3ziz replied:
this was just a workaround for the validation on `net.Dialer.DialContext`.
the services url we call are a mix of different schemes (https/wss/redis) but when i directly use the address on the dial it fails `dial tcp: address tfchain.grid.tf: missing port in address`.
so i am just adding the port 80 for all the address

muhamadazmy replied:
that's what i am saying. Dial because it's a tcp dial needs to know an ip and port. You can't assume that ALL services listening on port 80.  because a single IP can has multiple ports listening (multiple services).

For example say the hub. The hub has a web page that is available on port 80 and 443 (http and https) but also has zdb on port 9900. So if u assume port 80 for all your tests (if the port is not in the host name) then u will only be testing the http port.

Instead this function has to use the default port of the services. 

Luckily links with custom ports already have the port in the URL but your code has to be correct, and to be correct you need to use the default port **of the protocol** when guessing the port number

muhamadazmy commented on pkg/perf/networkhealth/networkhealth.go:
```diff
@@ -0,0 +1,112 @@
+	return host
+}
+
+func isReachable(ctx context.Context, serviceUrl string) error {
```
```suggestion
func isReachable(ctx context.Context, address string) error {
```

you call it serviceUrl while in fact it's an address of format `<host>:<port>`

Omarabdul3ziz committed:
resolve comments:
- use bus client instead of creating one in the diagnostics call
- get a connection instead of a pool for getting the perf result on diagnostics call
- separate the broker address parsing to reuse on NewRedisConn method

xmonader commented on client/node.go:
```diff
@@ -98,6 +99,19 @@ type Interface struct {
+	SystemStatusOk bool `json:"system_status_ok"`
+	// Modules is a list of modules with their objects and workers
+	Modules map[string]moduleStatus `json:"modules"`
+}
```
Why isn't this exported?

xmonader commented on pkg/perf/networkhealth/networkhealth.go:
```diff
@@ -0,0 +1,112 @@
+	"github.com/threefoldtech/zos/pkg/perf"
+)
+
+const requestTimeout = 5 * time.Second
```
if that at any point will be parametrized, make it `defaultRequestTimeout`

xmonader commented on pkg/perf/networkhealth/networkhealth.go:
```diff
@@ -0,0 +1,112 @@
+	servicesUrl := []string{
+		env.ActivationURL, env.GraphQL, env.FlistURL,
+	}
+	servicesUrl = append(servicesUrl, env.SubstrateURL...)
```
can't these two lines be joined?

xmonader commented on pkg/utils/redis.go:
```diff
@@ -8,42 +8,58 @@ import (
+	Options []redis.DialOption
+}
 
+func parseRedisAdd(address string) (RedisConfig, error) {
```
Rename to parseRedisAddress, why did you abbreviate that?

xmonader replied:
Also add tests for that

xmonader commented on pkg/utils/redis.go:
```diff
@@ -8,42 +8,58 @@ import (
-	} else if len(size) > 1 {
-		panic("invalid pool size")
-	}
+type RedisConfig struct {
```
This should be RedisDialParams

xmonader commented on client/node.go:
```diff
@@ -98,6 +99,19 @@ type Interface struct {
 	Mac string   `json:"mac"`
 }
 
+type moduleStatus struct {
```
Export it

xmonader commented on pkg/diagnostics/diagnostics.go:
```diff
@@ -0,0 +1,113 @@
+	"qsfsd",
+}
+
+type moduleStatus struct {
```
Why isn't that exported and reused in client/node.go?

Omarabdul3ziz committed:
refactors:
- export diagnostics types to be imported on node client
- rename to params
- add tests

muhamadazmy commented on pkg/perf/networkhealth/networkhealth.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+func (t *NetworkHealthTask) Jitter() uint32 {
+	return 0
```
we need to add jitter here or it means ALL nodes in the world will try to test the services at exactly the same time, DDoS(ing) our own services!

muhamadazmy replied:
I think this should be the number of seconds (check docs) but if we gonna do a test every 15 min the jitter could be around 5 min or so

muhamadazmy commented on pkg/perf/networkhealth/networkhealth.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+func (t *NetworkHealthTask) Cron() string {
+	return "0 */5 * * * *"
```
is 5 min a requirement, may be each 15 min is better

Omarabdul3ziz committed:
resolve comments:
- increase the schedule to be 15m
- add 5m jitter to not spam the services
- assign default port based on the url scheme

Omarabdul3ziz committed:
efficient use of redis pool:
- use redis pool saved in state instead of creating new connection each time for the diagnostics call
- assume node is offline if there is no network test yet
